### ISSUE TITLE
Better management of operator configuration changes

### ISSUE THREAD

**Issue by linsun** [istio/istio#27494]
**Describe the feature request**

Some of our configurations have been moved around while we don't offer any error to operators who is deploying IOP into Istio.  This has caused some issues for some users because the IOP isn't functioning properly as it appears to function.

One example is our team was using the following config to set outbound traffic policy defaults:

```
    global:
      outboundTrafficPolicy:
        mode: "ALLOW_ANY"
```
And it has been working fine in 1.5 + 1.6, e.g. https://github.com/istio/istio/blob/release-1.6/manifests/charts/global.yaml#L421

However, this field is no longer in [global.yaml](https://github.com/istio/istio/blob/release-1.7/manifests/charts/global.yaml) in 1.7 code base.  We learned this via **a very hard way** that this is moved and hard to update the config to:

```
  meshConfig:
    outboundTrafficPolicy: "ALLOW_ANY"
```

Trying to think how we can provide a better experience to our users (mainly operators role here):
1) Note these types of configuration moves in [update notice](https://istio.io/latest/news/releases/1.7.x/announcing-1.7/upgrade-notes/) or [release note](https://istio.io/latest/news/releases/1.7.x/announcing-1.7/change-notes/).
2) when a field is no longer supported, operator currently ignores the old configure we had (under global.outboundTrafficPolicy).  It would be so much better if we warn or even fail the IOP (for new apply) so users would know right away.

[ ] Docs
[x ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**

**Comment by linsun** [comment]
cc @kramvan1 @rvennam @sdake @ostromart @costinm @howardjohn

**Comment by howardjohn** [comment]
I am a bit confused. Not only does it work, we do have a deprecation notice - on both 1.6 and 1.7:
```
$ ~/istio/istio-1.7.0/bin/istioctl manifest generate --set values.global.outboundTrafficPolicy.mode=REGISTRY_ONLY | rg REGI
! values.global.outboundTrafficPolicy is deprecated; use meshConfig.outboundTrafficPolicy instead
                    - REGISTRY_ONLY
                    - REGISTRY_ONLY
          "mode": "REGISTRY_ONLY"
$ ~/istio/istio-1.6.8/bin/istioctl manifest generate --set values.global.outboundTrafficPolicy.mode=REGISTRY_ONLY | rg REGI
! global.outboundTrafficPolicy is deprecated; use meshConfig.outboundTrafficPolicy instead
                  - REGISTRY_ONLY
      mode: REGISTRY_ONLY
          "mode": "REGISTRY_ONLY"
```

**Comment by howardjohn** [comment]
And in master, it fails fast as requested:
```
$ istioctl manifest generate --set values.global.outboundTrafficPolicy.mode=REGISTRY_ONLY -d manifests
Run the command with the --force flag if you want to ignore the validation error and proceed.
Error: unknown field "outboundTrafficPolicy" in v1alpha1.GlobalConfig
```

Perhaps its a difference in istioctl vs in cluster operator?

**Comment by kramvan1** [comment]
@howard yes we are using the controller operator with an IOP resource.    I don't recall seeing deprecation messages in the operator logs, it just seems to ignore it and no errors.    
I wonder if the `istioctl manifest install --dry-run -f iop.yaml`  would show this error?
When I get a chance, I will try that.

**Comment by howardjohn** [comment]
Even with the lack of messages though, it should still be in 1.7 as far as I can tell from above? Are you running `master`?

**Comment by kramvan1** [comment]
This was for 1.7.0+ releases, and the global side of this stopped working/had no effect.  Only the new meshConfig field would take effect.

**Comment by sdake** [comment]
With istioctl 1.7.2:

```sdake@scale-cp:~$ ./istioctl manifest generate --set values.global.outboundTrafficPolicy.mode=ALLOW_ANY | grep REGI
! values.global.outboundTrafficPolicy is deprecated; use meshConfig.outboundTrafficPolicy instead
                    - REGISTRY_ONLY
                    - REGISTRY_ONLY
sdake@scale-cp:~$ ./istioctl manifest generate --set values.global.outboundTrafficPolicy.mode=ALLOW_ANY | grep ALLOW
! values.global.outboundTrafficPolicy is deprecated; use meshConfig.outboundTrafficPolicy instead
                - ALLOW
                    - ALLOW_ANY
                    - ALLOW_ANY
          "mode": "ALLOW_ANY"
```

THe bug report lists "allow_any" in lowercase. Perhaps that is the issue as allow_any is not an enum in the operator.

**Comment by rvennam** [comment]
This didn't work in 1.7.0:
```
  meshConfig:
    accessLogFile: /dev/stdout
    enablePrometheusMerge: true
    protocolDetectionTimeout: 0s
  profile: default
  values:
    global:
      logging:
        level: default:info
      outboundTrafficPolicy:
        mode: REGISTRY_ONLY
```

**Comment by howardjohn** [comment]
Ah ok - my grep was wrong. So 1.7 it gives a soft deprecation warning and does nothing. In master it fails hard. In 1.6 it works and has a deprecation warning.

Desired behavior here is in 1.7 it fails hard

**Comment by linsun** [comment]
Yes, just tried with istioctl 1.7.0 with a simple yaml from https://github.com/istio/istio/issues/27494#issuecomment-698032348... it did give a soft warning which is very easy to be ignored... I think the warning is wrong, should be the warning for 1.6... 1.7 is not deprecated, it is removed so should fail hard.

```
$ istioctl manifest generate -f test-iop.yaml                              
! values.global.outboundTrafficPolicy is deprecated; use meshConfig.outboundTrafficPolicy instead
kind: CustomResourceDefinition
apiVersion: apiextensions.k8s.io/v1
```

Probably worthwhile to see what other config options are like this and cherry pick some fail-hard fixes from master to 1.7 for options we no longer support.

**Comment by linsun** [comment]
> With istioctl 1.7.2:
> 
> ```
> sdake@scale-cp:~$ ./istioctl manifest generate --set values.global.outboundTrafficPolicy.mode=allow_any | grep allow
> Run the command with the --force flag if you want to ignore the validation error and proceed.
> Error: unknown value "allow_any" for enum v1alpha1.OutboundTrafficPolicyConfig_Mode
> 
> sdake@scale-cp:~$ ./istioctl manifest generate --set values.global.outboundTrafficPolicy.mode=ALLOW_ANY | grep ALLOW
> ! values.global.outboundTrafficPolicy is deprecated; use meshConfig.outboundTrafficPolicy instead
>                 - ALLOW
>                     - ALLOW_ANY
>                     - ALLOW_ANY
>           "mode": "ALLOW_ANY"
> ```
> 
> THe bug report lists "allow_any" in lowercase. Perhaps that is the issue as allow_any is not an enum in the operator.

Sorry that was just a typo.  Use REGISTRY_ONLY can see the prob fast.

**Comment by linsun** [comment]
cc @houshengbo

**Comment by morvencao** [comment]
Tried with in-cluster operator, same behavior with `istioctl`:
1. for 1.7, it doesn't block the installation with `global.outboundTrafficPolicy.mode. REGISTRY_ONLY` in iop, the difference is that no warning message in operator logs.
2. for 1.8, it will show error message `unknown field "outboundTrafficPolicy" in v1alpha1.GlobalConfig` and will not install istio.

The root cause is that we removed `outboundTrafficPolicy` from values proto(https://github.com/istio/istio/blob/master/operator/pkg/apis/istio/v1alpha1/values_types.proto) after 1.7.x. So the unmarshal will failed with unknown field.

If we really the same behavior in 1.7.x, we should remove the field `outboundTrafficPolicy` from values proto(https://github.com/istio/istio/blob/release-1.7/operator/pkg/apis/istio/v1alpha1/values_types.proto).
